text/0000-grand-testing-unification.md

```diff
@@ -0,0 +1,551 @@
+
+### Example of Proposed Syntax - QUnit
+
+#### Acceptance Tests
+
+```js
+import { moduleForAcceptance } from '../helpers/module-for-acceptance';
+import { test } from 'ember-qunit';
+
+moduleForAcceptance('Can see things');
+
+test('clicking on a thing redirects to other-thing', async function(assert) {
+  await this.visit('/things');
+
+  let list = this.find('.thing-item');
+  assert.equal(list.length, 5);
+
+  await this.click('.thing-item[data-id=1]');
+
+  assert.equal(this.currentRouteName, 'other-thing');
+});
+```
+
+#### Integration Tests
+
+```js
+import { moduleForIntegration } from '../helpers/module-for-integration';
+import { test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+
+moduleForIntegration('post-display');
+
+test('expands when clicked', async function(assert) {
+  await this.render(hbs`{{post-display}}`);
+
+  assert.equal(this.find('.post-created-at').length, 0, 'precond - no detaisl present');
+
+  await this.click('.post-item');
+
+  assert.equal(this.find('.post-created-at').textContent, '2016-01-05');
+});
+```
+
+# Migration Plan
+
+Since this is a fairly large departure from our current testing solution, a dedicated migration plan section seems appropriate.
+
+As discussed in the Detailed Design section, I am proposing to introduce new API's that do not currently exist in `ember-test-helpers`, `ember-qunit`, or `ember-mocha`. Once this is done, and the API's have been tested in a number of applications, we will update  the default Ember CLI blueprints to use the new structure and rewrite the guides.emberjs.com testing section. This will allow newly generated tests to use the new API's without immediately breaking existing test suites.
+
+Based on feedback received from the community on the usability of the new structure proposed here, we will deprecate usage of existing Ember API's:
+
+* `Ember.Test.registerHelper`
+* `Ember.Test.registerAsyncHelper`
+* `Ember.Application#setupForTesting`
+* `Ember.Test.unregisterHelper`
+* `Ember.Test.onInjectHelpers`
+* `Ember.Application#injectTestHelpers`
+* `Ember.Application#removeTestHelpers`
+* `Ember.Test.registerWaiters`
+* `Ember.Test.unregisterWaiters`
+* `Ember.Test.*`
+* existing `ember-qunit` / `ember-mocha` API's.
+
+It is possible that we can automate some of the migration process, we should continue to investigate migration automation as we progress with implementation.
+
+#### Example Migration of Component Integration Test
+
+This is [an existing test](https://github.com/aptible/dashboard.aptible.com/blob/09c76ef827704c065140fe890ef1a4d24c42be36/tests/integration/components/primitive-select-test.js) for a very simple `{{object-select}}` component:
+
+```js
+import Ember from 'ember';
+import { moduleForIntegration } from 'ember-qunit';
```

Comment:
`moduleForIntegration` is correct. There is nothing specific to components about it. It can be used to test helpers, or interactions between components. Think of it like a micro acceptance test.

On the other hand, unit testing a component, can be done with moduleForUnit, though I recommend sticking with integration tests unless you have a specific reason to unit test.